src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -375,6 +374,7 @@ public bool TryGetCompilation([NotNullWhen(true)] out Compilation? compilation)
                 if (state.FinalCompilationWithGeneratedDocuments != null && state.FinalCompilationWithGeneratedDocuments.TryGetValue(out var compilationOpt) && compilationOpt.HasValue)
                 {
                     compilation = compilationOpt.Value;
+                    state.InitializeUnrootedSymbolSet(compilation);
```

Comment:
i don't love that we expose compilations in effectively two separate codepaths (TryGetCompilation and GetCompilationAsync), and i need to update both.  feels fragile.

Reply:
I'm worried there's also potentially a third place now: CompilationTracker.GetPartialMetadataReference also hands out compilations which I guess get stitched into P2P references if you're freezing some consuming project. I'm _highly_ suspicious of that code now that I'm aware it exists because I suspect you can get compilations that hadn't called RecordAssemblySymbols on it _either_ and that might be bad in it's own right.

I _think_ a slightly more defensive way to write this (if this is indeed broken) is update RecordAssemblySymbols to also take a State parameter, and that method should update the the unrooted symbols for that compilation. The existing calls to RecordAssemblySymbols are updated with the states they have, and we also update GetPartialMetadataReference as well...I think.